components/org.wso2.carbon.identity.image/src/main/java/org/wso2/carbon/identity/image/file/FileBasedStorageSystemImpl.java

```diff
@@ -0,0 +1,223 @@
+import org.wso2.carbon.identity.image.exception.StorageSystemException;
+import org.wso2.carbon.identity.image.util.StorageSystemUtil;
+
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.channels.Channels;
+import java.nio.channels.FileChannel;
+import java.nio.channels.ReadableByteChannel;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.attribute.FileTime;
+
+import static org.wso2.carbon.identity.image.util.StorageSystemConstants.ID_SEPERATOR;
+import static org.wso2.carbon.identity.image.util.StorageSystemConstants.IMAGE_STORE;
+import static org.wso2.carbon.identity.image.util.StorageSystemConstants.SYSTEM_PROPERTY_CARBON_HOME;
+
+/**
+ * This is the implementation class to store images in to local file system.
+ */
+public class FileBasedStorageSystemImpl implements StorageSystem {
+
+    private static final Logger LOGGER = Logger.getLogger(FileBasedStorageSystemFactory.class);
+
+    @Override
+    public String addFile(InputStream inputStream, String type, String uuid, String tenantDomain)
+            throws StorageSystemException {
+
+        try {
+            return uploadImageUsingChannels(inputStream, type, uuid, tenantDomain);
+        } catch (IOException e) {
+            String errorMsg = String.format("Error while uploading image to file system for %s type.", type);
+            throw new StorageSystemException(errorMsg, e);
+        }
+
+    }
+
+    @Override
+    public InputStream getFile(String id, String type, String tenantDomain) throws StorageSystemException {
+
+        String[] urlElements = retrieveUrlElements(id);
+        InputStream inputStream;
+        try {
+            inputStream = getImageFile(urlElements, type, tenantDomain);
+        } catch (IOException e) {
+            String errorMsg = String.format("Error while retrieving the stored file of type %s.", type);
+            throw new StorageSystemException(errorMsg, e);
+        }
+        return inputStream;
+    }
+
+    @Override
+    public void deleteFile(String id, String type, String tenantDomain) throws StorageSystemException {
+
+        String[] urlElements = retrieveUrlElements(id);
+        try {
+            deleteImageFile(urlElements, type, tenantDomain);
+        } catch (IOException e) {
+            String errorMsg = String.format("Error while deleting the stored file of type %s.", type);
+            throw new StorageSystemException(errorMsg, e);
+        }
+
+    }
+
+    private String uploadImageUsingChannels(InputStream fileInputStream, String type, String uuid, String tenantDomain)
+            throws IOException {
+
+        int tenantId = IdentityTenantUtil.getTenantId(tenantDomain);
+        Path imagesPath = createSpecificDirectory(type, tenantId);
+        Path targetLocation = imagesPath.resolve(uuid);
+        try (FileOutputStream fileOutputStream = new FileOutputStream(targetLocation.toFile());
```

Comment:
Need more debug logs. 
e.g. writing file X to location Y
Byte count written
Success, and any exceptions, e.g Disk Full, file permission, read only state.